app/services/hotkeys.ts

```diff
@@ -504,46 +561,86 @@ export class Hotkey implements IHotkey {
     this.shouldApply = this.action.shouldApply(entityId);
   }
 
-
-  isSameHotkey(other: IHotkey) {
-    return (this.actionName === other.actionName) &&
-      (this.sceneId === other.sceneId) &&
-      (this.sourceId === other.sourceId) &&
-      (this.sceneItemId === other.sceneItemId);
+  equals(other: IHotkey) {
+    return (
+      this.actionName === other.actionName &&
+      this.sceneId === other.sceneId &&
+      this.sourceId === other.sourceId &&
+      this.sceneItemId === other.sceneItemId
+    );
   }
 
-
   getModel(): IHotkey {
     return { ...this.hotkeyModel };
   }
 
-
   private getAction(entityId: string): IHotkeyAction {
-    const action = { ...HOTKEY_ACTIONS[this.type].find(action => {
-      return action.name === this.actionName;
-    }) };
+    const action = getActionFromName(this.actionName);
 
     const { up, down } = action;
 
     // Fill in optional functions with some defaults
-    if (!action.isActive) action.isActive = () => false;
-    if (!action.shouldApply) action.shouldApply = () => true;
-
+    if (!action.isActive) {
+      action.isActive = () => false;
+    }
+    if (!action.shouldApply) {
+      action.shouldApply = () => true;
+    }
 
     // We defer the actions until after we've decided whether
     // or not to execute each action.
     if (up) {
       action.upHandler = () => {
-        if (!action.isActive(entityId)) defer(() => up(entityId));
+        if (!action.isActive(entityId)) {
+          defer(() => up(entityId, this.hotkeyModel.hotkeyId));
+        }
       };
     }
 
     if (down) {
       action.downHandler = () => {
-        if (!action.isActive(entityId)) defer(() => down(entityId));
+        if (!action.isActive(entityId)) {
+          defer(() => down(entityId, this.hotkeyModel.hotkeyId));
+        }
       };
     }
 
     return action;
   }
 }
+
+const getMigrationMapping = (actionName: string) => {
+  return {
+    MUTE: 'TOGGLE_MUTE',
+    UNMUTE: 'TOGGLE_UNMUTE',
+  }[normalizeActionName(actionName)];
+};
+
+const getActionFromName = (actionName: string) => {
+  return ACTIONS[actionName] || ACTIONS[getMigrationMapping(actionName)];
+};
+
+const isSupportedHotkey = (hotkey: OBSHotkey) =>
+  hotkey.ObjectType === obs.EHotkeyObjectType.Source &&
+  getActionFromName(hotkey.HotkeyName) &&
+  idPropFor(hotkey);
+
+const isSceneItem = (hotkey: OBSHotkey) => !!getScenesService().getSceneItem(hotkey.ObjectName);
+
+const isSource = (hotkey: OBSHotkey) => !!getSourcesService().getSource(hotkey.ObjectName);
+
+const isScene = (hotkey: OBSHotkey) => !!getScenesService().getScene(hotkey.ObjectName);
+
+const idPropFor = (hotkey: OBSHotkey) => {
+  if (isSource(hotkey)) {
+    return 'sourceId';
+  } else if (isScene(hotkey)) {
+    return 'sceneId';
+  } else if (isSceneItem(hotkey)) {
+    return 'sceneItemId';
+  } else {
+    return null;
+  }
+};
+
+const normalizeActionName = (actionName: string) => actionName.split('.')[0];
```

Comment:
i feel like utility functions should be defined above the class, we already have several examples of that in this file. open to putting them after but either way i think it should be consistent.

Reply:
I agree 100% with this, however I've mixed feelings about it. There wasn't an easy choice, as I believe the most important thing for us looking at this file would be
Types -> Consts -> Class -> Utils
Unfortunately, the functions you see above Class are there because of no-use-before-define. The rest is supporting and I believe should stay out of the way at the end. Unless we really want them in our way before we get to read the class, that is.